### PR TITLE
[Iceworks 3.x] 项目统一规范

### PR DESCRIPTION
**Pull request by alvinhui**
### 工程及项目操作解耦

- [Project adapter](https://github.com/alibaba/ice/blob/iceworks/project-stand/packages/iceworks-server/src/interface/project.ts#L269)
- [ice-scripts project adapter 示例](https://github.com/alibaba/ice/blob/iceworks/project-stand/packages/iceworks-server/src/lib/adapter/project.ts)

示例：

0. [前端发起“开启调试”请求](https://github.com/alibaba/ice/blob/iceworks/project-stand/packages/iceworks-client/src/pages/Dev/index.js#L20)
0. [服务端路由匹配](https://github.com/alibaba/ice/blob/iceworks/project-stand/packages/iceworks-server/src/app/router.ts#L4)
0. [服务端控制器进行响应](https://github.com/alibaba/ice/blob/iceworks/project-stand/packages/iceworks-server/src/app/io/controller/project.ts#L6)
0. [项目操作器负责调度](https://github.com/alibaba/ice/blob/iceworks/project-stand/packages/iceworks-server/src/lib/plugin/project-client/app.ts#L35)
0. [具体的项目 adapter 负责启动调试服务](https://github.com/alibaba/ice/blob/iceworks/project-stand/packages/iceworks-server/src/lib/adapter/project.ts#L77)

### 项目面板各插件的封装

以“页面列表”插件为例。

前端：

- [页面列表 component](https://github.com/alibaba/ice/blob/iceworks/project-stand/packages/iceworks-client/src/pages/Project/components/Page/index.js)
- [页面列表 store](https://github.com/alibaba/ice/blob/iceworks/project-stand/packages/iceworks-client/src/pages/Project/stores/pages.js)

服务端：

- [页面列表 controller](https://github.com/alibaba/ice/blob/iceworks/project-stand/packages/iceworks-server/src/app/controller/api.ts#L34)
- [页面列表 service](https://github.com/alibaba/ice/blob/iceworks/project-stand/packages/iceworks-server/src/lib/service/page.ts)

## 项目相关能力

- [x] 现有功能：
    - [x] 页面管理
        - [x] 页面列表
            - view: [/renderer/src/pages/Home/ProjectDashboard/Pages/index.jsx](https://github.com/alibaba/ice/blob/iceworks%2Frelease-3.x/tools/iceworks/renderer/src/pages/Home/ProjectDashboard/Pages/index.jsx#L164)
            - data: [recursivePagesSync](https://github.com/alibaba/ice/blob/iceworks%2Frelease-3.x/tools/iceworks/renderer/src/pages/Home/ProjectDashboard/Pages/index.jsx#L34)
        - [x] 添加页面（存在消息通信：状态同步）
            - view: 
                - 添加区块：[/renderer/src/pages/Home/CreatPage/index.jsx](https://github.com/alibaba/ice/blob/iceworks%2Frelease-3.x/tools/iceworks/renderer/src/pages/Home/CreatPage/index.jsx#L101)
                - 填写信息：[/renderer/src/pages/Home/CreatPage/PageConfig/index.js](https://github.com/alibaba/ice/blob/iceworks%2Frelease-3.x/tools/iceworks/renderer/src/pages/Home/CreatPage/PageConfig/index.js#L340)
            - data: [/app/main/scaffolder/lib/createPage.js](https://github.com/alibaba/ice/blob/iceworks%2Frelease-3.x/tools/iceworks/app/main/scaffolder/lib/createPage.js#L38)
        - [x] 添加区块（存在消息通信：状态同步）
            - view: [/renderer/src/components/PageBlockPicker/index.jsx](https://github.com/alibaba/ice/blob/iceworks%2Frelease-3.x/tools/iceworks/renderer/src/components/PageBlockPicker/index.jsx#L188)
            - data: [downloadBlocksToPage](https://github.com/alibaba/ice/blob/iceworks%2Frelease-3.x/tools/iceworks/app/main/scaffolder/lib/utils.js#L28)
        - [x] 删除页面
            - view: [handlePageDelete](https://github.com/alibaba/ice/blob/iceworks%2Frelease-3.x/tools/iceworks/renderer/src/pages/Home/ProjectDashboard/Pages/index.jsx#L130)
            - data: [/app/main/scaffolder/lib/removePage.js](https://github.com/alibaba/ice/blob/iceworks%2Frelease-3.x/tools/iceworks/app/main/scaffolder/lib/removePage.js)
    - [x] 布局管理
        - [x] 布局列表
    - [x] 依赖管理
        - [x] 依赖列表
        - [x] 添加依赖（存在消息通信：日志输出）
            - view: [/renderer/src/pages/Home/ProjectPanel/AddPackage.jsx](https://github.com/alibaba/ice/blob/iceworks%2Frelease-3.x/tools/iceworks/renderer/src/pages/Home/ProjectPanel/AddPackage.jsx)
            - data: [project-scripts/npminstall](https://github.com/alibaba/ice/blob/iceworks%2Frelease-3.x/tools/iceworks/renderer/src/lib/project-scripts.js#L338)
        - [x] 升级依赖
    - [x] TODO
- [x] 新增功能：
    - [x] 菜单管理
        - [x] 菜单列表
        - [x] 添加菜单
        - [x] 删除菜单
        - [x] 更新菜单
    - [x] 路由管理
        - [x] 路由列表
        - [x] 添加路由
        - [x] 删除路由
        - [x] 更新路由
    - [x] Mock 管理
        - [x] Mock 列表
        - [x] 添加 Mock
        - [x] 删除 Mock
        - [x] 更新 Mock
    - [x] 物料信息
        - [x] 区块列表
        - [x] 组件列表
            - [x] 升级组件
        - [x] 模板信息
- [ ] 其他
    - [ ] Git
    - [ ] DEF
    - [ ] OSS
- [x] 工程相关
    - [x] start
    - [x] stop
    - [x] build

## 2.x Terminal 现状

前端渲染：

- [/src/pages/Home/ProjectPanel](https://github.com/alibaba/ice/blob/iceworks/release-3.x/tools/iceworks/renderer/src/pages/Home/ProjectPanel/index.jsx)：项目面板
- [/src/components/ProjectTerminal](https://github.com/alibaba/ice/blob/iceworks/release-3.x/tools/iceworks/renderer/src/components/ProjectTerminal/index.jsx)：项目的终端
- [/src/terms](https://github.com/alibaba/ice/blob/iceworks/release-3.x/tools/iceworks/renderer/src/terms/index.js)：终端管理器
- [/src/spc](https://github.com/alibaba/ice/tree/iceworks/release-3.x/tools/iceworks/renderer/src/spc.js)：终端事件封装

命令执行：

- [/renderer/src/lib/project-scripts.js](https://github.com/alibaba/ice/blob/iceworks%2Frelease-3.x/tools/iceworks/renderer/src/lib/project-scripts.js)：封装项目相关操作，如 start/build/stop/install
- [/app/main/services/sessions/manager.js](https://github.com/alibaba/ice/blob/iceworks%2Frelease-3.x/tools/iceworks/app/main/services/sessions/manager.js)
- [/app/main/services/sessions/Session.js](https://github.com/alibaba/ice/blob/iceworks%2Frelease-3.x/tools/iceworks/app/main/services/sessions/Session.js)
- [/app/main/services/sessions/startProxy.js](https://github.com/alibaba/ice/blob/iceworks%2Frelease-3.x/tools/iceworks/app/main/services/sessions/startProxy.js)
- [/app/main/services/sessions/buildProxy.js](https://github.com/alibaba/ice/blob/iceworks%2Frelease-3.x/tools/iceworks/app/main/services/sessions/buildProxy.js)

## 2.x 主进程与渲染进程通信

渲染进程的事件监听：

- `focus`: 界面获得系统的焦点，即 iceworks 成为活动窗口时触发；
- `menu:project:console:toggle`：切换运行日志窗口，由程序的菜单栏（项目 - 运行日志）触发；
- `router.push`：跳去某个页面，由程序的菜单栏（iceworks - 设置）触发；
- `updater-message`：程序需要更新，由自动升级检查触发；
- `update_project`：更新项目信息，由项目启动调试的流程触发（告知渲染进程当前项目调试的状态：构建中/编译完成/运行中）；
- `menu:project:clearall`：删除所有项目，由程序的菜单栏（项目 - 清空列表）触发；
- `processTracking`：进度条中打印状态，由创建页面的程序触发；
- `progressVisible`：是否显示进度条，由创建页面的程序触发；
- `session:data`：启动的子进程任务有日志输出（将会往 iceworks 的终端面板写入数据），由各种命令操作触发；
- `session:exit`：启动的子进程退出（iceworks 的终端面板将会新起一行）。

### CODE REVIEW

File: packages/iceworks-client/src/components/XtermTerminal/index.js [hunk 1]
```diff
@@ -1,31 +1,40 @@
...
+import { ICEWORKS_PROJECT_DEV_DATA } from 'iceworks-events';
 import { useSocket } from '@hooks/useSocket';
 import * as fit from 'xterm/dist/addons/fit/fit';
 import * as webLinks from 'xterm/dist/addons/webLinks/webLinks';
+import stores from '@stores';
 import 'xterm/dist/xterm.css';
+import log from '@utils/logger';
 import styles from './index.module.scss';
 
+const logger = log.getLogger('xterm');
```
Comment (chenbin92) [inline packages/iceworks-client/src/components/XtermTerminal/index.js]: 这里不太语义，看到其他地方引用如下：

image

改为：

```
import logger from '@utils/logger';

const log = logger.getLogger('xterm');
```

File: packages/iceworks-client/src/pages/Project/components/Dependency/index.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import React from 'react';
+import stores from '../../stores';
```
Comment (chenbin92) [inline packages/iceworks-client/src/pages/Project/components/Dependency/index.js]: 搞个 `alias`

File: packages/iceworks-client/src/pages/Project/components/Page/index.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
+import React from 'react';
+import stores from '../../stores';
```
Comment (chenbin92) [inline packages/iceworks-client/src/pages/Project/components/Page/index.js]: use alias

File: packages/iceworks-client/src/pages/Project/index.js [hunk 4]
```diff
@@ -1,14 +1,24 @@
 import React, { useEffect } from 'react';
 import { Input, Button } from '@alifd/next';
 import stores from '@stores';
+import logger from '@utils/logger';
+import Page from './components/Page';
+import Dependency from './components/Dependency';
```
Comment (chenbin92) [inline packages/iceworks-client/src/pages/Project/index.js]: Rename Dependencies

File: packages/iceworks-client/src/pages/Project/index.js [hunk 5]
```diff
@@ -1,14 +1,24 @@
 import React, { useEffect } from 'react';
 import { Input, Button } from '@alifd/next';
 import stores from '@stores';
+import logger from '@utils/logger';
+import Page from './components/Page';
```
Comment (chenbin92) [inline packages/iceworks-client/src/pages/Project/index.js]: Rename Pages

File: packages/iceworks-client/src/stores/currentProject.js [hunk 6]
```diff
@@ -1,37 +1,46 @@
...
     await new Promise(resolve => setTimeout(resolve, 1000));
 
     const { pages } = this.dataSource;
     this.pages = [].concat(pages).concat([{ ...page, id: pages.length }]);
   },
-  async setData(project) {
-    await new Promise(resolve => setTimeout(resolve, 1000));
-    this.dataSource = project;
+  async reset(folderPath) {
+    const response = await fetch(`${appConfig.apiUrl}project/current`, {
```
Comment (chenbin92) [inline packages/iceworks-client/src/stores/currentProject.js]: 统一通过 socket 处理

File: packages/iceworks-server/src/app/controller/api.ts [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+  pageService: IPluginService;
+
+  @get('/')
+  async getAll(ctx) {
+    ctx.body = await this.pageService.getAll(ctx.query.projectFolderPath);
+  }
+}
+
+@provide()
+@controller('/api/dependency')
```
Comment (chenbin92) [inline packages/iceworks-server/src/app/controller/api.ts]: Rename Dependencies

File: packages/iceworks-server/src/app/io/controller/project.ts [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+      let error;
+      let project
+      try {
+        project = await projectClient.devStart(projectFolderPath);
+      } catch(err) {
+        error = err;
+      }
+
+      if (project) {
+        project.on('dev:data', function(data) {
```
Comment (chenbin92) [inline packages/iceworks-server/src/app/io/controller/project.ts]: EventName 统一大写？

File: packages/iceworks-server/src/lib/adapter/project.ts [hunk 9]
```diff
@@ -0,0 +1,123 @@
+import * as path from 'path';
+import * as pathExists from 'path-exists';
+import * as fs from 'fs';
+import junk from 'junk';
+import * as util from 'util';
+import * as child_process from 'child_process';
+import * as detectPort from 'detect-port';
+import * as EventEmitter from 'events';
+import { IProjectPage, IProjectDependency } from '../../interface';
+const originalReaddir = util.promisify(fs.readdir);
```
Comment (chenbin92) [inline packages/iceworks-server/src/lib/adapter/project.ts]: `originalReaddir`  命名为` readdirAsync` 会更语义一点

File: packages/iceworks-server/src/lib/adapter/project.ts [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+
+const readdir = async (targetPath) => {
+  if (pathExists.sync(targetPath)) {
+    return (await originalReaddir(targetPath)).filter(junk.not);
+  }
+  return [];
+};
+
+const recursive = async function(dirPath) {
```
Comment (chenbin92) [inline packages/iceworks-server/src/lib/adapter/project.ts]: 这个 fn 做的事情好像并不是递归读取文件，只是读了 pages 下的文件？

直接用这个 https://www.npmjs.com/package/recursive-readdir

File: packages/iceworks-server/src/lib/adapter/project.ts [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+  async getPages(): Promise<IProjectPage[]> {
+    return recursive(path.join(this.folderPath, 'src', 'pages'));
+  }
+
+  async getDependencies(): Promise<IProjectDependency[]> {
+    return [
+      {
+        package: 'icestore',
+        dev: false,
+        specifyingVersion: '^0.1.0'
```
Comment (chenbin92) [inline packages/iceworks-server/src/lib/adapter/project.ts]: 这个为什么要单独获取？

File: packages/iceworks-server/src/lib/service/dependency.ts [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+
+@provide('dependencyService')
+export class DependencyService implements IPluginService {
+    @plugin('projectClient')
+    private projectClient;
+
+    async getAll(projectFolderPath: string): Promise<IPluginGetAllResult> {
+        const project = this.projectClient.getProject(projectFolderPath);
+        return {
+            data: project ? await project.getDependencies() : []
```
Comment (chenbin92) [inline packages/iceworks-server/src/lib/service/dependency.ts]: 统一两个空格，格式化一下

File: packages/iceworks-server/src/lib/storage.ts [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+  delete(key: string): void {
+    this.store.delete(key);
+  }
+}
+
+const schema = {
+  'stat': {
+    type: 'string',
+  },
+  'iceland-stat': {
```
Comment (chenbin92) [inline packages/iceworks-server/src/lib/storage.ts]: `iceland-stat` 是啥意思？

File: packages/iceworks-server/src/lib/storage.ts [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+  'projects': {
+    type: 'array',
+    items: {
+      type: 'string',
+    },
+    default: []
+  }
+};
+
+export default new Store({schema});
```
Comment (chenbin92) [inline packages/iceworks-server/src/lib/storage.ts]: 可以直接用社区的？
https://www.npmjs.com/package/node-storage
https://www.npmjs.com/package/node-localstorage

File: packages/iceworks-server/src/interface/project.ts [hunk 15]
```diff
@@ -0,0 +1,527 @@
...
+export interface IProjectRouter {
+  /**
+   * URL 路径
+   */
+  path: string;
+
+  /**
+   * 页面名
+   */
+  pageName: string;
```
Comment (chenbin92) [inline packages/iceworks-server/src/interface/project.ts]: 直接代码这里写的是 component，是命名伟 pageName 还是 component ？

image

File: packages/iceworks-server/src/interface/project.ts [hunk 16]
```diff
@@ -0,0 +1,527 @@
...
+export interface IProjectDependency {
+  /**
+   * 包名
+   */
+  package: string;
+
+  /**
+   * 指定版本：^1.0.1 / ~1.0.1 / 0.0.x
+   */
+  specifyingVersion: string;
```
Comment (chenbin92) [inline packages/iceworks-server/src/interface/project.ts]: 应该是 `specifyVersion`?

File: packages/iceworks-server/src/interface/project.ts [hunk 17]
```diff
@@ -0,0 +1,527 @@
...
+     * 文本
+     */
+    text: string;
+  }>;
+}
+
+/**
+ * 项目的依赖信息
+ */
+export interface IProjectDependency {
```
Comment (chenbin92) [inline packages/iceworks-server/src/interface/project.ts]: 这个看起来约定的作用是？直接读取项目的 package.json 就可以了？

File: packages/iceworks-server/src/interface/project.ts [hunk 18]
```diff
@@ -0,0 +1,527 @@
...
+  /**
+   * 背景图
+   */
+  image?: string;
+}
+
+/**
+ * TODO 项目的布局
+ */
+export interface IProjectLayout {
```
Comment (chenbin92) [inline packages/iceworks-server/src/interface/project.ts]: 这个应该没有了

File: packages/iceworks-server/src/lib/service/dependency.ts [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+    }
+
+    async create() {
+
+    }
+
+    async delete() {
+
+    }
+}
```
Comment (chenbin92) [inline packages/iceworks-server/src/lib/service/dependency.ts]: 知道 service 为什么放在 lib 下吗？

image

File: packages/iceworks-server/src/lib/plugin/project-client/app.ts [hunk 20]
```diff
@@ -0,0 +1,60 @@
+import * as EventEmitter from 'events';
+import storage from '../../storage';
+import Project from '../../adapter/project';
+import getEnv from '../../getEnv';
+
+class ProjectsClient {
```
Comment (chenbin92) [inline packages/iceworks-server/src/lib/plugin/project-client/app.ts]: `ProjectsClient` 应该改为 `ProjectClient`? 文件路径也是 `plugin/project-client`

Comment (chenbin92) [reply packages/iceworks-server/src/lib/plugin/project-client/app.ts]: 另外 ProjectsClient 怎么理解？

File: packages/iceworks-server/src/lib/adapter/project.ts [hunk 21]
```diff
@@ -0,0 +1,123 @@
...
+    const { folderPath } = this;
+    const env = { PORT: port };
+
+    if (this.devProcess) {
+      throw new Error('调试服务已启动，不能多次启动，请先停止已启动的调试服务后再次启动');
+    }
+
+    const childProcess = child_process.spawn('npm', ['start'], { 
+      cwd: folderPath,
+      env: Object.assign({}, setEnv, env)
```
Comment (chenbin92) [inline packages/iceworks-server/src/lib/adapter/project.ts]: `setEnv` 语义是 fn 需要被调用，但实上是一个对象， `setEnv: any` 也不太语义

File: packages/iceworks-client/src/components/XtermTerminal/index.js (hunk 1 above)
Comment (alvinhui) [reply packages/iceworks-client/src/components/XtermTerminal/index.js]: 这里关键是统一声明处还是统一调用处。我选择了统一调动处。
即，所有代码内都是 `logger.info`。

File: packages/iceworks-client/src/pages/Project/components/Dependency/index.js (hunk 2 above)
Comment (alvinhui) [reply packages/iceworks-client/src/pages/Project/components/Dependency/index.js]: 页面级别也要搞？

File: packages/iceworks-server/src/app/io/controller/project.ts (hunk 8 above)
Comment (alvinhui) [reply packages/iceworks-server/src/app/io/controller/project.ts]: 目前是全部小写，包括 socket API 部分

Comment (alvinhui) [reply packages/iceworks-server/src/app/io/controller/project.ts]: 统一使用了小数点做分割

File: packages/iceworks-client/src/pages/Project/components/Dependency/index.js (hunk 2 above)
Comment (chenbin92) [reply packages/iceworks-client/src/pages/Project/components/Dependency/index.js]: 页面级别就近原则放置不需要啊，`src/store` 搞一个即可？

File: packages/iceworks-server/src/lib/adapter/project.ts (hunk 11 above)
Comment (alvinhui) [reply packages/iceworks-server/src/lib/adapter/project.ts]: 这是一个示例数据

File: packages/iceworks-server/src/lib/storage.ts (hunk 13 above)
Comment (alvinhui) [reply packages/iceworks-server/src/lib/storage.ts]: 这是从老数据中直接拿过来的，我先删除掉。

File: packages/iceworks-server/src/lib/plugin/project-client/app.ts (hunk 20 above)
Comment (alvinhui) [reply packages/iceworks-server/src/lib/plugin/project-client/app.ts]: 1. 已改名
2. 实际上是个调度器，可以看看职责想个其他名字。Projects 我觉得不太友好，无法从名字理解它的用途。

File: packages/iceworks-server/src/lib/adapter/project.ts (hunk 21 above)
Comment (alvinhui) [reply packages/iceworks-server/src/lib/adapter/project.ts]: rename to settingsEnv

File: packages/iceworks-server/src/lib/adapter/project.ts [hunk 22]
```diff
@@ -0,0 +1,123 @@
...
+
+  private devProcess: child_process.ChildProcess;
+
+  public devStatus: string = DEV_STATUS_NORMAL;
+
+  constructor(folderPath: string) {
+    super();
+    this.name = path.basename(folderPath);
+    this.folderPath = folderPath;
+  }
```
Comment (chenbin92) [inline packages/iceworks-server/src/lib/adapter/project.ts]: 可以改为这样？
```
name: path.basename(folderPath),
folderPath: folderPath
```

**Pull request merged**